### PR TITLE
feat: register first unread message id on message.new & when initializing Channel state

### PR DESCRIPTION
**Pull request by MartinCupela**
## Description of the changes, What, Why and How?

To achieve consistency in the UI concerning the display of unread messages indicators, we need to provide the `first_unread_message_id` value consistently. The value is not provided when querying channels, neither when a new message arrives. Therefore, we have to calculate it client-side.
In this PR we register the `first_unread_message_id` in `message.new` handler and in `Channel._initializeState()`. The latter is invoked every time a `Channel` instance is created or channel data is queried from the BE. In `Channel._initializeState()` we inspect the whole message set and not only the recently queried messages.

This PR also removes dead code, that was anyways overwritten on the following lines.

### CODE REVIEW

**Review by szuperaz — COMMENTED** [review]
Can't we use this algorithm:https://www.notion.so/stream-wiki/Mark-Unread-Developer-Specs-3b05e4fb069a4e2b9355216a5e6085a2?pvs=4#04bc470a33e249c6ba9e7e6c8d6c9279? That algorithm computes `first_unread_message_id` based on `last_read_message_id` so we would have to run that algorithm whenever the `last_read_message_id` can change, which are these places based on the code:
- `message.read` event
- `notification.mark_unread` event
- `_initializeState` function

**Comment by MartinCupela** [comment]
> Can't we use this algorithm:https://www.notion.so/stream-wiki/Mark-Unread-Developer-Specs-3b05e4fb069a4e2b9355216a5e6085a2?pvs=4#04bc470a33e249c6ba9e7e6c8d6c9279? That algorithm computes `first_unread_message_id` based on `last_read_message_id` so we would have to run that algorithm whenever the `last_read_message_id` can change, which are these places based on the code:
> 
> * `message.read` event
> * `notification.mark_unread` event
> * `_initializeState` function

I have tested that `last_read_message_id` is not guaranteed to be provided. I actually had an implementation with that one, but as it does not have to be always available I decided to go the safe path.

**Comment by szuperaz** [comment]
> > Can't we use this algorithm:https://www.notion.so/stream-wiki/Mark-Unread-Developer-Specs-3b05e4fb069a4e2b9355216a5e6085a2?pvs=4#04bc470a33e249c6ba9e7e6c8d6c9279? That algorithm computes `first_unread_message_id` based on `last_read_message_id` so we would have to run that algorithm whenever the `last_read_message_id` can change, which are these places based on the code:
> > 
> > * `message.read` event
> > * `notification.mark_unread` event
> > * `_initializeState` function
> 
> I have tested that `last_read_message_id` is not guaranteed to be provided. I actually had an implementation with that one, but as it does not have to be always available I decided to go the safe path.

Did you find a scenario where the implementation based on `last_read_message_id` wasn't working? I'm asking this because iOS uses that algorithm ([code](https://github.com/GetStream/stream-chat-swift/blob/develop/Sources/StreamChat/Controllers/ChannelController/ChannelController.swift#L1220)) and mark unread feature is already available there. So if the code works for them, it can work for us too. It feels safer to copy an algorithm already used in an SDK, than to create a new one.

It also feels more consistent, even though it's possible your algorithm does the same as iOS but it's hard to compare the two .

**Comment by MartinCupela** [comment]
I 

> > > Can't we use this algorithm:https://www.notion.so/stream-wiki/Mark-Unread-Developer-Specs-3b05e4fb069a4e2b9355216a5e6085a2?pvs=4#04bc470a33e249c6ba9e7e6c8d6c9279? That algorithm computes `first_unread_message_id` based on `last_read_message_id` so we would have to run that algorithm whenever the `last_read_message_id` can change, which are these places based on the code:
> > > 
> > > * `message.read` event
> > > * `notification.mark_unread` event
> > > * `_initializeState` function
> > 
> > 
> > I have tested that `last_read_message_id` is not guaranteed to be provided. I actually had an implementation with that one, but as it does not have to be always available I decided to go the safe path.
> 
> Did you find a scenario where the implementation based on `last_read_message_id` wasn't working? I'm asking this because iOS uses that algorithm ([code](https://github.com/GetStream/stream-chat-swift/blob/develop/Sources/StreamChat/Controllers/ChannelController/ChannelController.swift#L1220)) and mark unread feature is already available there. So if the code works for them, it can work for us too. It feels safer to copy an algorithm already used in an SDK, than to create a new one.
> 
> It also feels more consistent, even though it's possible your algorithm does the same as iOS but it's hard to compare the two .

I personally have some read states in queried channels without `last_read_message_id` + the type is declared with it as optional whereas `last_read` is mandatory.

image

**Comment by szuperaz** [comment]
> I
> 
> > > > Can't we use this algorithm:https://www.notion.so/stream-wiki/Mark-Unread-Developer-Specs-3b05e4fb069a4e2b9355216a5e6085a2?pvs=4#04bc470a33e249c6ba9e7e6c8d6c9279? That algorithm computes `first_unread_message_id` based on `last_read_message_id` so we would have to run that algorithm whenever the `last_read_message_id` can change, which are these places based on the code:
> > > > 
> > > > * `message.read` event
> > > > * `notification.mark_unread` event
> > > > * `_initializeState` function
> > > 
> > > 
> > > I have tested that `last_read_message_id` is not guaranteed to be provided. I actually had an implementation with that one, but as it does not have to be always available I decided to go the safe path.
> > 
> > 
> > Did you find a scenario where the implementation based on `last_read_message_id` wasn't working? I'm asking this because iOS uses that algorithm ([code](https://github.com/GetStream/stream-chat-swift/blob/develop/Sources/StreamChat/Controllers/ChannelController/ChannelController.swift#L1220)) and mark unread feature is already available there. So if the code works for them, it can work for us too. It feels safer to copy an algorithm already used in an SDK, than to create a new one.
> > It also feels more consistent, even though it's possible your algorithm does the same as iOS but it's hard to compare the two .
> 
> I personally have some read states in queried channels without `last_read_message_id` + the type is declared with it as optional whereas `last_read` is mandatory.
> 
> image

But this is expected, and the iOS algorithm handles it (`Scenario 3`). I think the `last_read_message_id` is missing in this specific example because you haven't opened the channel since the `last_read_message_id` field has been implemented.

**Comment by MartinCupela** [comment]
> > I
> > > > > Can't we use this algorithm:https://www.notion.so/stream-wiki/Mark-Unread-Developer-Specs-3b05e4fb069a4e2b9355216a5e6085a2?pvs=4#04bc470a33e249c6ba9e7e6c8d6c9279? That algorithm computes `first_unread_message_id` based on `last_read_message_id` so we would have to run that algorithm whenever the `last_read_message_id` can change, which are these places based on the code:
> > > > > 
> > > > > * `message.read` event
> > > > > * `notification.mark_unread` event
> > > > > * `_initializeState` function
> > > > 
> > > > 
> > > > I have tested that `last_read_message_id` is not guaranteed to be provided. I actually had an implementation with that one, but as it does not have to be always available I decided to go the safe path.
> > > 
> > > 
> > > Did you find a scenario where the implementation based on `last_read_message_id` wasn't working? I'm asking this because iOS uses that algorithm ([code](https://github.com/GetStream/stream-chat-swift/blob/develop/Sources/StreamChat/Controllers/ChannelController/ChannelController.swift#L1220)) and mark unread feature is already available there. So if the code works for them, it can work for us too. It feels safer to copy an algorithm already used in an SDK, than to create a new one.
> > > It also feels more consistent, even though it's possible your algorithm does the same as iOS but it's hard to compare the two .
> > 
> > 
> > I personally have some read states in queried channels without `last_read_message_id` + the type is declared with it as optional whereas `last_read` is mandatory.
> > image
> 
> But this is expected, and the iOS algorithm handles it (`Scenario 3`). I think the `last_read_message_id` is missing in this specific example because you haven't opened the channel since the `last_read_message_id` field has been implemented.

So you are suggesting to use `last_read` as a fallback in case `last_read_message_id` does not exist?

**Comment by szuperaz** [comment]
> > > I
> > > > > > Can't we use this algorithm:https://www.notion.so/stream-wiki/Mark-Unread-Developer-Specs-3b05e4fb069a4e2b9355216a5e6085a2?pvs=4#04bc470a33e249c6ba9e7e6c8d6c9279? That algorithm computes `first_unread_message_id` based on `last_read_message_id` so we would have to run that algorithm whenever the `last_read_message_id` can change, which are these places based on the code:
> > > > > > 
> > > > > > * `message.read` event
> > > > > > * `notification.mark_unread` event
> > > > > > * `_initializeState` function
> > > > > 
> > > > > 
> > > > > I have tested that `last_read_message_id` is not guaranteed to be provided. I actually had an implementation with that one, but as it does not have to be always available I decided to go the safe path.
> > > > 
> > > > 
> > > > Did you find a scenario where the implementation based on `last_read_message_id` wasn't working? I'm asking this because iOS uses that algorithm ([code](https://github.com/GetStream/stream-chat-swift/blob/develop/Sources/StreamChat/Controllers/ChannelController/ChannelController.swift#L1220)) and mark unread feature is already available there. So if the code works for them, it can work for us too. It feels safer to copy an algorithm already used in an SDK, than to create a new one.
> > > > It also feels more consistent, even though it's possible your algorithm does the same as iOS but it's hard to compare the two .
> > > 
> > > 
> > > I personally have some read states in queried channels without `last_read_message_id` + the type is declared with it as optional whereas `last_read` is mandatory.
> > > image
> > 
> > 
> > But this is expected, and the iOS algorithm handles it (`Scenario 3`). I think the `last_read_message_id` is missing in this specific example because you haven't opened the channel since the `last_read_message_id` field has been implemented.
> 
> So you are suggesting to use `last_read` as a fallback in case `last_read_message_id` does not exist?

No, I'm suggesting to use the algorithm that iOS has implemented. In the case `last_read_message_id` is not defined it uses `Scenario 3`: 

Screenshot 2024-01-19 at 09 53 24

I'm trying to understand why we aren't using the algorithm specified in the Notion document. You mentioned the edge case of `last_read_message_id` not being defined, but the iOS algorithm handles that specific edge case (see above).

**Comment by MartinCupela** [comment]
> > > > I
> > > > > > > Can't we use this algorithm:https://www.notion.so/stream-wiki/Mark-Unread-Developer-Specs-3b05e4fb069a4e2b9355216a5e6085a2?pvs=4#04bc470a33e249c6ba9e7e6c8d6c9279? That algorithm computes `first_unread_message_id` based on `last_read_message_id` so we would have to run that algorithm whenever the `last_read_message_id` can change, which are these places based on the code:
> > > > > > > 
> > > > > > > * `message.read` event
> > > > > > > * `notification.mark_unread` event
> > > > > > > * `_initializeState` function
> > > > > > 
> > > > > > 
> > > > > > I have tested that `last_read_message_id` is not guaranteed to be provided. I actually had an implementation with that one, but as it does not have to be always available I decided to go the safe path.
> > > > > 
> > > > > 
> > > > > Did you find a scenario where the implementation based on `last_read_message_id` wasn't working? I'm asking this because iOS uses that algorithm ([code](https://github.com/GetStream/stream-chat-swift/blob/develop/Sources/StreamChat/Controllers/ChannelController/ChannelController.swift#L1220)) and mark unread feature is already available there. So if the code works for them, it can work for us too. It feels safer to copy an algorithm already used in an SDK, than to create a new one.
> > > > > It also feels more consistent, even though it's possible your algorithm does the same as iOS but it's hard to compare the two .
> > > > 
> > > > 
> > > > I personally have some read states in queried channels without `last_read_message_id` + the type is declared with it as optional whereas `last_read` is mandatory.
> > > > image
> > > 
> > > 
> > > But this is expected, and the iOS algorithm handles it (`Scenario 3`). I think the `last_read_message_id` is missing in this specific example because you haven't opened the channel since the `last_read_message_id` field has been implemented.
> > 
> > 
> > So you are suggesting to use `last_read` as a fallback in case `last_read_message_id` does not exist?
> 
> No, I'm suggesting to use the algorithm that iOS has implemented. In the case `last_read_message_id` is not defined it uses `Scenario 3`:
> 
> Screenshot 2024-01-19 at 09 53 24
> 
> I'm trying to understand why we aren't using the algorithm specified in the Notion document. You mentioned the edge case of `last_read_message_id` not being defined, but the iOS algorithm handles that specific edge case (see above).

Ok, let me adjust it

File: src/channel.ts [hunk 1]
```diff
@@ -1547,6 +1535,17 @@ export class Channel<StreamChatGenerics extends ExtendableGenerics = DefaultGene
         };
 
         if (read.user.id === user?.id) {
+          if (this.state.read[user.id].last_read) {
```
Comment (szuperaz) [inline src/channel.ts:1538]: Can't we only run this code if the underlying data has changed?

File: src/channel.ts [hunk 2]
```diff
@@ -1547,6 +1535,17 @@ export class Channel<StreamChatGenerics extends ExtendableGenerics = DefaultGene
...
+            for (let i = 0; i < messageSet.messages.length; i++) {
```
Comment (szuperaz) [inline src/channel.ts:1539]: `findLast` only supported since node v18, nice catch @MartinCupela

File: src/channel.ts [hunk 3]
```diff
@@ -1300,12 +1299,13 @@ export class Channel<StreamChatGenerics extends ExtendableGenerics = DefaultGene
               } else {
                 channelState.read[userId].unread_messages += 1;
               }
+              if (userId === ownUserId && channelState.read[ownUserId].unread_messages === 1) {
```
Comment (szuperaz) [inline src/channel.ts:1302]: Can you please add a comment here @MartinCupela ? It's a bit hard to understand

File: src/channel.ts (hunk 1 above)
Comment (MartinCupela) [reply src/channel.ts]: Fixed and added tests

File: src/channel.ts (hunk 3 above)
Comment (MartinCupela) [reply src/channel.ts:1303]: Added a comment.

**Review by szuperaz — APPROVED** [review]

**Comment by MartinCupela** [comment]
It will be unnecessary to calculate first unread message id. If not available at the moment of jumping to the first unread message, it will be determined as a message following the last read message.

**Pull request closed without merging**